js/pages/prediction/prediction-browser.js

```diff
@@ -0,0 +1,120 @@
+define([
+	'knockout',
+	'text!./prediction-browser.html',
+	'appConfig',
+	'./const',
+	'webapi/MomentAPI',
+	'webapi/AuthAPI',
+	'providers/Page',
+	'providers/AutoBind',
+	'utils/CommonUtils',
+	'services/Prediction',
+	'components/cohortcomparison/ComparativeCohortAnalysis',
+	'faceted-datatable',
+	'components/ac-access-denied',
+	'components/heading',
+	'components/empty-state',
+	'less!./prediction-browser.less'
+], function(
+	ko,
+	view,
+	config,
+	constants,
+	momentApi,
+	authApi,
+	Page,
+	AutoBind,
+	commonUtils,
+	PredictionService,
+) {
+  class PredictionBrowser extends AutoBind(Page) {
+		constructor(params) {
+			super(params);
+			this.model = params.model;
+			this.reference = ko.observableArray();
+			this.loading = ko.observable(true);
+
+			this.canReadPredictions = ko.pureComputed(() => {
+				return (config.userAuthenticationEnabled && authApi.isAuthenticated() && authApi.isPermittedReadPlps()) || !config.userAuthenticationEnabled;
+			});
+			this.canCreatePrediction = ko.pureComputed(() => {
+				return (config.userAuthenticationEnabled && authApi.isAuthenticated() && authApi.isPermittedCreatePlps()) || !config.userAuthenticationEnabled;
+			});
+
+			this.options = {
+				Facets: [
+					{
+						'caption': 'Last Modified',
+						'binding': function (o) {
```

Comment:
Doesn't copy-pasting this block of code all over the place tell that it should be moved into a shared function? Anyway, that was done, pls use `DatatableUtils.getFacetForDate`

Reply:
'all over the place' is a bit of hyperbole.  This utility class was just merged into master 2 weeks ago, and, I don't know about you, when I copy in some code, I don't go ahead and do a full codebase search to see if there is a utility class that performs the work.  I think we all need to come to an understanding that when something is added to the codebase, it isn't automatically encoded into the brains of everyone developing code (especially when much of this code is done in parallel.  

I think the underlying issue here is that we don't have an effective mechanism for introducing new efficiencies in the codebase in a way that people are aware of when and how to use it.  This is the case with BEMHelper + Less. This is the case with the Page life cycle.  This is the case for the new utility class functions.  

We have the Atlas/WebAPI calls to bring awareness of these things to the community. We've been having them for several weeks now, and no one brought up these items to raise awareness.  So, for the next few WebAPI calls, let's make sure that we raise these items for awareness...bring up the code, show how they are used in practice.

Reply:
A bit, yes. But:
- the construction was definitely copy-pasted from some other place; so even at this point it would be logical to move it into an util method and re-use
- it was copied at least twice in the PR (hard to see precise - too many not expanded files); the second time where I would personally think of moving copy-pasted stuff into a shared place

> I think the underlying issue here is that we don't have an effective mechanism for introducing new efficiencies

That's true, no doubts. Neither Page, nor BEMHelper are easy to get w/o intro that's why you haven't seen me mentioning them in the review. But LESS, being around for a long while, I thought is a bit more straightforward thing.

Reply:
Great.  absolutely agree that if it looks like we're copying things, we should think about encapsulation and re-use. That is a fair assessment.

But let's agree that it's fair to call out areas of where we could be leveraging technology/frameworks where they aren't use in practice, and that if we aren't using them, it's a symptom that it needs to be addressed at the Atlas/WebAPI calls and instruction on it's use.

Let's also agree that until that communication occurs, no one is 'on the hook' for doing it 'correctly'.  That's why it would be unfair to criticize your team for not following existing patterns in the code base (which we have never gave you formal introduction to the code) while at the same time it _is_ unfair to criticize design choices have been made without understanding the context and reasoning behind those decisions.  I'm sure we have a lot to learn from each other from our respective experiences.  So let's enter into these debates in the spirit of mutual understanding.

Reply:
> Let's also agree that until that communication occurs, no one is 'on the hook' for doing it 'correctly'

I thought we've discussed UI refactoring all together and received a lot of valid feedback from you, guys, same as you reviewed our PRs so that I assumed the communication took place (which is not the same thing with WebAPI, where we haven't done such broad communication yet). Anyway, this comment was about at least 3 blocks of the exactly same code blocks, so isn't it valid? And I just pointed that currently the function already exists so no need to code it from the ground.

Reply:
> Anyway, this comment was about at least 3 blocks of the exactly same code blocks, so isn't it valid? 

Yes, it's valid statement, which is where I started with:

> Great. absolutely agree that if it looks like we're copying things, we should think about encapsulation and re-use. That is a fair assessment.

Where it got on to the 'communication' topic was that we should have known that there was some util function that we should all be using for that column. That led into the other thread, but I think we're all on the same page now.